### ISSUE TITLE
Using Docker, Meilisearch v0.24 does not write data in /data.ms anymore

### ISSUE THREAD

**Issue by sovanna** [meilisearch/MeiliSearch#1969]
**Describe the bug**
Following the docs in Getting Started / Installation / Download and launch / Docker, it says that Meilisearch write data in /data.ms. 
When new user try it with Docker using persistence volume, they will use `-v $(pwd)/data.ms:/data.ms`.
But looking inside the folder, they will find no data.

**To Reproduce**
1. Go to [https://docs.meilisearch.com/learn/getting_started/installation.html#download-and-launch](https://docs.meilisearch.com/learn/getting_started/installation.html#download-and-launch)
2. Click on 'Docker' tab
3. Copy/Paste the command to start using meilisearch
4. Check your `$(pwd)/data.ms` folder on your host machine
5. No data is persist

**Expected behavior**
Data appears in the right folder of the host machine

**MeiliSearch version: v0.24**

**Additional context**
When changing user [https://github.com/meilisearch/MeiliSearch/issues/1757](https://github.com/meilisearch/MeiliSearch/issues/1757) the **workspace dir** has also changed. It is no more `/data.ms` but `/home/meili/data.ms`

We could update the documentation or remove the `WORDIR ${HOME}` and let it be in `/data.ms`

**Comment by curquiza** [comment]
Hello @sovanna, thank you so much for this report! 🙏 sorry for the breaking change we did not anticipate enough regarding the docs!

I will create a PR to update the documentation today :)

Also, I opened a more general discussion about the `data.ms` path: https://github.com/meilisearch/product/discussions/332

I will close the issue once it's fixed in the docs.

**Comment by curquiza** [comment]
Here is the PR in the docs: https://github.com/meilisearch/documentation/pull/1329

The [changelogs of v0.24.0](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.24.0) have been updated accordingly

**Comment by curquiza** [comment]
The docs are now up to date! Thanks again for your report! 🙏

**Comment by sanders41** [comment]
Using `/home/meili/data.ms` results in an error: `Error: An internal error has occurred. 'Permission denied (os error 13)'.`

```sh
docker run -it --rm \
    -p 7700:7700 \
    -v $(pwd)/data.ms:/home/meili/data.ms \
    getmeili/meilisearch:latest
```

**Comment by sovanna** [comment]
@sanders41 indeed

Well, on MacOS (12.0.1), with Docker Desktop 4.3.0, the command works fine.
However, I've just tried to run the same command on a Debian 11 and I had the same error as you 😢

I wanted to deploy a docker-compose stack with a meilisearch service but I got stuck.
The *quick fix* 😱  I've found was to `chown` the host folder with the same UID:GID as meili user.

The user `meili` created by the Dockerfile has `UID=1000` et `GID=1000`.

I've read that you can also pass those settings as ARG in the Dockerfile but it would mean that we have to handle our own Dockerfile.
Others have tried with entrypoint.sh that adapt the UID et GID of the user inside the container with the host.
@see [https://github.com/sudo-bmitch/jenkins-docker/blob/main/entrypoint.sh](https://github.com/sudo-bmitch/jenkins-docker/blob/main/entrypoint.sh)

The weird thing is that on MacOS, I don't have a user with UID=1000 and GID=1000 and it works.
I don't really understand why on MacOS I don't have those permissions issues 😅

**Comment by sanders41** [comment]
If I change the user to `root` in the docker-compose file it works without error. Not a good solution though. Maybe docker desktop on the Mac is automatically running as root? That is the only thing I can think of.

For reference I am using Manjaro so it seems as though any Linux distro has the issue.

**Comment by sovanna** [comment]
aa more weird than ever,

Running on Debian 11 as root 
```
docker run -it --rm \
    -p 7700:7700 \
    -v $(pwd)/data.ms:/home/meili/data.ms \
    getmeili/meilisearch:latest
```
give me the error too.

It creates the `data.ms` on the host system.
Then I `chown -R 1000:1000 data.ms`
and re-run and `MeiliSearch` appears

same for root or non-root

**Comment by curquiza** [comment]
Docker enforces using it with the Root right on MacOs if I'm correct.

@sovanna Is the permission error a problem you only have with v0.24.0? Do you have the same error using v0.23.0 or before (so before we did the change related to the `meili` user)?

**Comment by sanders41** [comment]
@sovanna I have the same issue running as root on the host side. Instead I had to set the docker user to root. Overall I think I like your solution of setting the permissions on the host side better. The down side is someone has to remember to do that anywhere the container is run rather than having it be automatic.

```sh
docker run -it --rm \
    -p 7700:7700 \
    -u root \
    -v $(pwd)/data.ms:/home/meili/data.ms \
    getmeili/meilisearch:latest

**Comment by sovanna** [comment]
On **Debian 11 Server**
with **Docker version 20.10.11, build dea9396**

As a non-root user logged in to the server,

## 1. With meilisearch:latest aka v0.24.0 (/home/meili/data.ms)
```
docker run -it --rm \
    -p 7700:7700 \
    -v $(pwd)/data.ms:/home/meili/data.ms \
    getmeili/meilisearch:latest
```
Got 
```
Error: An internal error has occurred. `Permission denied (os error 13)`.
```


## 2. With meilisearch:latest aka v0.23.0 (/data.ms)
```
docker run -it --rm \
    -p 7700:7700 \
    -v $(pwd)/data.ms:/data.ms \
    getmeili/meilisearch:v0.23.0
```
Works!


## 3. As @sanders41 mentionned (using -u root or -u 0:0) for v0.24.0
```
docker run -it --rm \
    -p 7700:7700 \
    -u 0:0 \
    -v $(pwd)/data.ms:/home/meili/data.ms \
    getmeili/meilisearch:latest
```
Works!


## 4. Chown mounted folder to 1000:1000
```
mkdir -p $(pwd)/data.ms
chown -R 1000:1000 $(pwd)/data.ms
docker run -it --rm \
    -p 7700:7700 \
    -v $(pwd)/data.ms:/home/meili/data.ms \
    getmeili/meilisearch:latest
```
Works!

## 5. Using -u 1000:1000 (because meili user inside has uid=1000, gid=1000)
```
docker run -it --rm \
    -p 7700:7700 \
    -u 1000:1000 \
    -v $(pwd)/data.ms:/home/meili/data.ms \
    getmeili/meilisearch:latest
```
Got 
```
Error: An internal error has occurred. `Permission denied (os error 13)`.
```

---

For v0.24.0 (meili user), **case 3 and 4 work** with *extra*

**Comment by FLevent29** [comment]
Same on Windows 11.
I downgraded to `0.23.0`, because I have no idea how to set permissions in a Docker volume with docker-compose.

```yaml
version: "3.1"
services:
  MeiliSearch:
    image: getmeili/meilisearch:v0.23.0
    ports:
      - "7700:7700"
    volumes:
      - "meili:/home/meili/data.ms"
    environment:
      MEILI_MASTER_KEY: "masterKey"
volumes:
  meili:
```

However that comes with it's own problems, like
```typescript
await this.meili.getOrCreateIndex('records', { primaryKey: 'rec_id' });
```
throws `MeiliSearchApiError: Index records not found`, which if I understand correctly shouldn't happen, and in `0.24.0` works properly.

I guess I'll run MeiliSearch `0.24.0` without Docker until this gets fixed.

**Comment by FLevent29** [comment]
I made `./docker/Dockerfile-meili`
```dockerfile
FROM getmeili/meilisearch:v0.24.0
USER root
```
`./docker-compose.yaml`
```yaml
version: "3.1"
services:
  meilisearch:
    build:
      dockerfile: "Dockerfile-meili"
      context: "docker"
    ports:
      - "7700:7700"
    volumes:
      - "meili:/home/meili/data.ms"
    environment:
      MEILI_MASTER_KEY: "masterKey"
volumes:
  meili:
```
This will do for now.

**Comment by sanders41** [comment]
@FLevent29 if it is easier you can uses your `docker-compose.yaml` file for this instead of a separate `Dockerfile`.

```yaml
version: "3.1"
services:
  meilisearch:
    image: getmeili/meilisearch:v0.24.0
    user: root
    ports:
      - "7700:7700"
    volumes:
      - "meili:/home/meili/data.ms"
    environment:
      MEILI_MASTER_KEY: "masterKey"
volumes:
  meili:
```

As a side note, you probably want to read your `MEILI_MASTER_KEY` from an environment variable rather than hard coding it in `MEILI_MASTER_KEY: "masterKey"` because if you commit it to your repository people will be able to see it.

**Comment by sanders41** [comment]
Maybe adding a step in the `Dockerfile` something like `chmod 666 /home/meili` before changing to the `meili` user would allow the volumes to work without having to run as `root`? I’m not sure if it will work or not. If not going back to `root` being the default user is the only other solution I can think of.

**Comment by sanders41** [comment]
@K2ouMais does your docker-compose use `image: getmeili/meilisearch:v0.24.0`? If yes that image is built with the `Dockerfile` we are discussing so a fix there would also fix on your compose side.

**Comment by K2ouMais** [comment]
Unfortunately I am using the standard docker-compose file that comes with Laravel Sail.

It has this line:
```image: getmeili/meilisearch:latest```

Yesterday I tried it with your solution and it didn't work.

With that I mean that the data couldn't be found in the volume.

**Comment by sanders41** [comment]
Currently `image: getmeili/meilisearch:latest` is the same as `image: getmeili/meilisearch:v0.24.0` so if we come up with a solution it will fix both. The v0.25 release isn't far off so it's also possible the fix comes with that release, and would also cover `latest` since `latest` would bump to `v0.25.0`.

As you already found you won't have the issue with `image: getmeili/meilisearch:v0.23.0`.

**Comment by sanders41** [comment]
@curquiza I tried the `chmod` idea along with several other things and always end up with the permission denied error. I can't think of anything else to try other than going back to running as `root`.

**Comment by sovanna** [comment]
@sanders41 i've tried also to test something with the Dockerfile but it took me too much time to compile Rust... damn laptop..

Basically, the approach that I've tried was to create also the /home/meili/data.ms with the right permission and ownership but it failed because at some point, Meili try to look for a version file inside the /home/meili/data.ms when you create it prematurely.

The idea was to let `root` handle things like installation, permission, ownership etc.. and then finally, exec meilisearch as meili user with su-exec through the entrypoint.sh

**Comment by sanders41** [comment]
@sovanna I tried something similar without success. Looking closer at the entrypoint.sh, what I didn’t try was adjusting the docker group id.

**Comment by sovanna** [comment]
I tried the Dockerfile postgres approach or even redis as they all do the kind of same approach but failed miserably. At least, I learned a few things on bash script, on gosu or su-exec 🤣

Some others put their id user to 999 and I've tried too but the Dockerfile is based on Alpine and the user id was already taken.

I'll continue to search on it this weekend but I am also afraid to be blocked by how /data.ms is created and handled rust-side.
Because, have you tried @sanders41 to create your `data.ms` before started `meiliserch`? I did and it looks for a version file and if not failed to start.

**Comment by sanders41** [comment]
I have not tried to create the `data.ms` ahead of time. I just took a quick look through the code and I don’t think it is going to be possible to create the `data.ms` ahead of time, I believe MeiliSearch is going to need to create that itself because it looks to be doing several things.

**Comment by curquiza** [comment]
Hum, regarding the error you have and also since the core-team does not have time to make the dockerfile works with a `meili` user, I think we have to roll back this change introduced in v0.24.0. I'll let you know, but this problem will be fixed in v0.25.0.

**Comment by K2ouMais** [comment]
https://github.com/laravel/sail/pull/299

**Comment by caugner** [comment]
@sanders41 For me, the command runs without problems:

```terminal
$ cd /tmp
$ mkdir data.ms
$ docker run -it --rm \
    -p 7700:7700 \
    -v $(pwd)/data.ms:/home/meili/data.ms \
    getmeili/meilisearch:latest

888b     d888          d8b 888 d8b  .d8888b.                                    888
8888b   d8888          Y8P 888 Y8P d88P  Y88b                                   888
88888b.d88888              888     Y88b.                                        888
888Y88888P888  .d88b.  888 888 888  "Y888b.    .d88b.   8888b.  888d888 .d8888b 88888b.
888 Y888P 888 d8P  Y8b 888 888 888     "Y88b. d8P  Y8b     "88b 888P"  d88P"    888 "88b
888  Y8P  888 88888888 888 888 888       "888 88888888 .d888888 888    888      888  888
888   "   888 Y8b.     888 888 888 Y88b  d88P Y8b.     888  888 888    Y88b.    888  888
888       888  "Y8888  888 888 888  "Y8888P"   "Y8888  "Y888888 888     "Y8888P 888  888

Database path:          "./data.ms"
Server listening on:    "http://0.0.0.0:7700"
Environment:            "development"
Commit SHA:             "0a1d2ce2318c26f36ddcb82d3159b8003120b84f"
Commit date:            "2021-11-15T18:02:07+00:00"
Package version:        "0.24.0"

Thank you for using MeiliSearch!

We collect anonymized analytics to improve our product and your experience. To learn more, including how to turn off analytics, visit our dedicated documentation page: https://docs.meilisearch.com/learn/what_is_meilisearch/telemetry.html

Anonymous telemetry:    "Enabled"
Instance UID:           "36bee4e3-bc53-4352-af85-fae34b500cdd"

No master key found; The server will accept unidentified requests. If you need some protection in development mode, please export a key: export MEILI_MASTER_KEY=xxx

Documentation:          https://docs.meilisearch.com
Source code:            https://github.com/meilisearch/meilisearch
Contact:                https://docs.meilisearch.com/resources/contact.html or bonjour@meilisearch.com

[2021-12-15T08:56:10Z INFO  actix_server::builder] Starting 2 workers
[2021-12-15T08:56:10Z INFO  actix_server::builder] Starting "actix-web-service-0.0.0.0:7700" service on 0.0.0.0:7700
```

**Comment by caugner** [comment]
> i've tried also to test something with the Dockerfile but it took me too much time to build the docker image... damn laptop..

@sovanna Did you try to run `docker build --cache-from=getmeili/meilisearch:latest -t meili .`? This should save you a lot of build time, if you make the test changes somewhere near the end of the `Dockerfile`.

**Comment by curquiza** [comment]
FYI, @ManyTheFish will work on it, trying to find a fix, otherwise, we will rollback.
A RC of v0.25.0 will be released before 21th December, so that you can check it work if you have the time, before the official v0.25.0 release (10/01/2022)

**Comment by K2ouMais** [comment]
It would be really bad, if Laravel takes meilisearch out of sail because of all these problems.

Common guys... Fix it or revert it.

**Comment by curquiza** [comment]
Sorry, I would say revert instead of rollback sorry

**Comment by caugner** [comment]
@ManyTheFish I wonder whether a symlink from `/data.ms` to `/home/meili/data.ms` could be a backwards-compatible solution.

**Comment by K2ouMais** [comment]
I loged into the running meilisearch container:
```
docker exec -it --user root containerid /bin/sh
```

And symlinked eveything under ```/home/meili/data.ms``` in the data.ms root structure.
```
cd /
ln -sn /home/meili/data.ms/* data.ms
```

This way I can have all the folders under the data.ms folder.

I wonder if this could help to make the data persistent again.

**Comment by michaelbaudino** [comment]
FYI, I'm using the following wrapper image that works well to workaround this bug:

```Dockerfile
FROM getmeili/meilisearch:v0.24.0

USER root

RUN echo -e '#!/bin/sh\n\nchown -R meili ${HOME}\n\nexec "$@"' \
      > /usr/local/bin/chown-home-entrypoint && \
    chmod +x /usr/local/bin/chown-home-entrypoint

USER ${USER}

ENTRYPOINT ["chown-home-entrypoint", "tini", "--"]
CMD ./meilisearch
```

Would you like a PR with this change?

**Comment by ManyTheFish** [comment]
Hey all! I'm currently trying to fix this, but, because it's not a trivial fix, we will revert the PR that introduces this bug for the `v0.25` and rewrite a new Dockerfile in another PR for `v0.26`. 

Thanks for your feedbacks! 🙌

**Comment by michaelbaudino** [comment]
Alright, thanks for the update @ManyTheFish.

Is there anything that my "chown-ing directory in entrypoint" patch above wouldn't cover? (I'm new to Meilisearch, I probably don't see all implications)

**Comment by ManyTheFish** [comment]
Hi @michaelbaudino thanks,
I tried your solution and it to works.
However, I find it a bit patchy, and doesn't seems to be a "clean" fix of our issue. 😞

But, I found a [pretty long discussion about workarounds or features](https://github.com/moby/moby/issues/2259) to fix this. There are:
- "chown scripts" like yours
- make the user change the right of the dest volume to 1000:1000
- [using named volumes](https://github.com/moby/moby/issues/2259#issuecomment-916982087) which, I think, is a relatively "clean" workaround for development. ⚖️ 

Now @ all! (@michaelbaudino, @K2ouMais, @caugner, @sanders41, @sovanna, @FLevent29).
Is your issue:
- a development issue? 👍 (where named volumes could work to fix this issue)
```diff
docker run -it --rm \
    -p 7700:7700 \
-    -v $(pwd)/data.ms:/home/meili/data.ms \
+    -v meilisearch_db:/home/meili/data.ms \
    getmeili/meilisearch:latest
```
- a production issue? 👎 (where named volumes can't be a fix of this issue)

### warning ⚠️ 

the current `getmeili/meilisearch:latest` doesn't work with named volume too, but I can easily patch it to make it work.

if you want to try named volumes, use this dummy Dockerfile:
```bash
FROM    alpine:3.14

RUN adduser -D meili

RUN mkdir -p /home/meili/data.ms && \
    chown meili /home/meili/data.ms

USER meili

CMD     ["touch", "/home/meili/data.ms/database"]
```

build:
`docker build -t dummy_meili .`

run:
` docker run -it --rm -v meilisearch_db:/home/meili/data.ms dummy_meili:latest`

Don't hesitate to give me more feedbacks about this! 🙏

**Comment by sanders41** [comment]
@ManyTheFish for me it is currently development, but the plan is to move the same setup to production once things are ready.

**Comment by sovanna** [comment]
@ManyTheFish in development, I use bind mounts and volume for production.
At that time, I didn't know why it didn't work so I changed to bind mounts and changed the ownership.
I preferred updating ownership as It was more simple to do than adding a script or a Dockerfile.

For development, I am usually using bind mounts as I like to have everything in the project folder. Also, using docker volume for development is storing data on the Docker storage and "one time", Docker Desktop failed, thus I've lost everything... Sure, in dev, it's ok, data was fake but still frustrating to re-generate some data etc...

Generally speaking, my feedback on this would be that it should be transparent regarding using bind mounts or volume.

I am also a bit biased by what I've already read on that while investigating.
See [https://stackoverflow.com/questions/36781372/docker-using-gosu-vs-user](https://stackoverflow.com/questions/36781372/docker-using-gosu-vs-user)
So, I would say that @michaelbaudino solution is acceptable to me

**Comment by curquiza** [comment]
Sorry for the delay, [v0.25.0rc2](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.25.0rc2) is out, containing the docker fix: `data.ms` is now accessible from root (`/data.ms`) and not from `home` (`/home/meili/data.ms`) anymore

**Comment by fevernova90** [comment]
@curquiza thank you! Saved my day, using 0.24 and having the same issue and don't want to rebuild the image myself.

**Comment by alfonsrv** [comment]
Tried working around this (and future possible corruptions) using `MEILI_DB_PATH=/data` which then results in `Error: Version file is missing or the previous MeiliSearch engine version was below 0.24.0. Use a dump to update Meilisearch.`

If an empty directory is found, a new meilisearch-structure should ideally be created inside of it automatically.

**Comment by curquiza** [comment]
Hello @alfonsrv, I'm not sure to understand. What are your steps to get this error? Did you try with the latest release, I mean [v0.25.0rc2](https://github.com/meilisearch/MeiliSearch/releases/tag/v0.25.0rc2)?

**Comment by alfonsrv** [comment]
No, I'm currently using rc0, because I exhausted my Docker pulls. 😿 

Simply using a `docker-compose` file trying to hardcode the directory. `meili-data` is an empty/automatically created directory on the host-side. Since the [docs](https://docs.meilisearch.com/reference/features/configuration.html#database-path) do not specify any UID-requirement and the error-message does not indicate a permission error, I did not investigate further as it seemed to me that Meilisearch requires some *magic file* to indicate the index's current version in a Docker environment.

```yml
  meili:
    image: "getmeili/meilisearch:v0.25.0rc0"
    container_name: meilibae
    environment:
      - MEILI_ENV=production
      - MEILI_HTTP_ADDR=${MEILI_HOST:-meili}:${MEILI_PORT:-7700}
      - MEILI_NO_ANALYTICS=true
      - MEILI_MASTER_KEY=${MEILI_MASTER_KEY}
      - MEILI_HTTP_PAYLOAD_SIZE_LIMIT=209715200
      - MEILI_DB_PATH=/data
    ports:
      - ${MEILI_PORT:-7700}:7700
    volumes:
      - ./meili-data:/data
    restart: unless-stopped
```

**Comment by curquiza** [comment]
rc0 has indeed this issue, this is fixed using rc2 :)

**Comment by curquiza** [comment]
Closed in favor of https://github.com/meilisearch/MeiliSearch/issues/2051

For any other issues that are not related to the `meili` user we have in the dockerfile, please open a new one.
